[WIP] add connection pool class

nkeyes opened the pull request.

Commit:
add connection pool class

Commit:
fix connection conditional

Commit:
add comment  from HappyBase about making the first connection

jonm-okc commented on lib/ok_hbase/pool.rb:
```diff
@@ -0,0 +1,83 @@
+    end
+
+    def connection(timeout = nil)
+      connection = Thread.current[:current_connection]
```
Might want to use a symbol name like `:ok_hbase_connection`, just in case something else like ActiveRecord uses a symbol called `:current_connection`

nkeyes replied:
good call

jonm-okc commented on lib/ok_hbase/pool.rb:
```diff
@@ -0,0 +1,83 @@
+        return_after_use = true
+        connection = _acquire_connection(timeout)
+        @_lock.synchronize do
+          Thread.current[:current_connection] = connection
```
Why are you synchronizing this block?  `Thread.current[:current_connection]` would be distinct for each thread, so there isn't any sort of a resource conflict here.

I think it might be the line above (the `_acquire_connection`) that you want to make sure only one thread is doing at a time.

nkeyes replied:
This is an artifact of the Python code I started from. Python has a bug where acquiring the lock there is needed. I'll fix it.

jonm-okc commented on lib/ok_hbase/pool.rb:
```diff
@@ -0,0 +1,83 @@
+    end
+
+    def _return_connection(connection)
+      @_connection_queue << connection
```
Have you checked whether enqueueing an element like this is thread safe?  Or might you need to wrap that in a synchronize call?

nkeyes replied:
Yeah, the Queue class in ruby is meant specifically for inter-thread usage: http://www.ruby-doc.org/stdlib-1.9.3/libdoc/thread/rdoc/Queue.html

jonm-okc commented on lib/ok_hbase/pool.rb:
```diff
@@ -0,0 +1,83 @@
+      # The first connection is made immediately so that trivial
+      # mistakes like unresolvable host names are raised immediately.
+      # Subsequent connections are connected lazily.
+      self.connection {}
```
I think this is something that you should allow the app to decide to do.  In Phoenix, there may be instances where the thread pool gets created due to some initialization code, but we don't want to immediately try connection.  For example, it would be great if Phoenix could do a `rake -T` without attempting to connect to hbase.

nkeyes replied:
i'll add a flag to handle this

Commit:
only eagerly load the first connection if opts[:auto_connect] is truthy

Commit:
rename current thread connection variable name to be more specific

Commit:
remove unneeded mutex/lock

Commit:
Merge branch 'master' of github.com:okcwest/ok_hbase into connection-pool

Commit:
fix current connection reset

Commit:
require ok_hbase/pool

Commit:
add OkHbase::Connection#ping?

Commit:
handle connection failures better

Commit:
add Connection#reset

Commit:
track pool connection ids
Refactor:
connection -> with_connection
_acquire_connection -> get_connection
_return_connection -> return_connection

Commit:
Merge branch 'master' of github.com:okcwest/ok_hbase into connection-pool
